### DISCUSSION TITLE
How do you associate Airflow task instances with additional context?

### DISCUSSION THREAD

**Discussion by karakanb** [apache/airflow#26462] [Q&A] [answered]
I am integrating Airflow with an internal system, and one of the problems I am running into is to associate additional metadata with task instances.

For example, I am building various pipelines dynamically based on a configuration, and depending on various configuration values the tasks I generate change. When I look at a dag run or a task instance, I need to be able to add that additional context into the task instance so that I can visualize it based on the task instance itself, hence include versioned information by default, but currently it is not possible.

Ideally, I am looking for something that allows me to add arbitrary metadata to a Dag and a Task Instance, and fetch it via the rest API. Is there anyone that has a way of doing it currently?

**Comment by potiuk** [comment]
You can store the data in S3 or GCP (or similar) and name your file and name it like `{dag_id}_{task_id}_{dag_run}`. 
Super easy and you do not put extra load on airflow DB to store meta-data, and you can use S3/GCP (or similar) API to retrieve your meta-data.

Any problem with it ?

**Reply by karakanb** [reply]
Maybe I couldn't express the problem very well, but what I meant was the configuration could spawn out many tasks/operators of different kind; in that sense, what file do you mean naming and storing in S3? for example, a single configuration could result in a pipeline that has 5 BashOperator tasks, 2 Python tasks, and 3 BigQueryInsertJobOperator tasks; as far as I know, there is no way I can hook something into the task instance creation that'd allow me to put additional data somewhere, even to S3.

What I am looking for is: which task instance is built based on what configuration? ideally, I'd like to be able to associate additional metadata with the dagruns / task instances, in a way that says "config version c5e3f6b was used for this task instance", ideally with some additional metadata. If I am not missing something, such an association is not possible today, am I wrong?

**Reply by potiuk** [reply]
You can prepare any metadata when you Dynamically Create the DAG. There is only one place where you create a DAG (this is DAG file) and usually you generate dynamic tasks there. THEN you can save any data whereever you want. For example you can dump a file with your configuraiton parameters:

```
DAG_ID:TASK_ID_1:<DAG/TASK_1 METADATA>
DAG_ID:TASK_ID_2:<DAG/TASK_2 METADATA>
```

And store the file in S3. And pull/retrieve the file when you need it. And name the file as you want.

Or ou can store separate_file for each task if you mostly need to retrieve it for single task:

File1: `PATH/DAG_ID/TASK_ID_1`:
```
TASK_1_METADATA
```

File2: `PATH/DAG_ID/TASK_ID_2`:
```
TASK_2_METADATA
```

Then if you want to retrieve metaadata for TASK2 - you pull the file via s3 API from `PATH/DAG_ID/TASK_ID_2` and retrieve metadata for it. 

What's the problem with that approach? why do you need something else ?

**Reply by karakanb** [reply]
The problem with this approach is that it saves metadata on the dag serialization level, not on dagrun creation; which means, there is no way I can tell based on which configuration this dagrun was created from.

Allow me to give an example:

- configuration version 1 results in two tasks for dag1:
  - a bash operator that says "hello world"
  - a BigQueryExecuteOperator that runs "select 1" on a bq database
- configuration version 2 results in two tasks for dag1 as well:
  - a bash operator that says "what a world"
  - a BigQueryExecuteOperator that runs "select 2" on a bq database

In this setup, if I get the dagruns, I'll see bunch of dagruns that were created based on v1, and another bunch on v2; however, since there is no way I can attach any information that persists with the dagruns, there is no way I can tell if a dagrun was created based on v1 and v2. Similarly, a task instance might be created with a different configuration even inside the same dagrun, e.g. the configuration is updated during a dagrun or a backfill is executed, therefore this metadata is required also on task instance level for this case.

In reality, the version changes are happening much more frequently, and the resulting pipelines have complex relationships with the configurations, which means without having the ability to attach a metadata to the dagrun in some ways, I don't have the ability to associate pipeline with configurations.

The things I have thought of doing:
- trying to associate versions based on timestamps of version updates / task instance execution times: this is a very complex thing, and given that they are coming from different sources with different timings, it is not possible to associate them confidently.
- making the versioning / metadata as part of the dag / task name: this means the history of the dags would be lost after every change, not feasible.
- storing these configuration information as part of dag description / task docs: these are not persisted across dagruns / task instances, so not a possibility.
- storing this information as part of an empty task included in the pipeline: this is the most promising one so far, this could allow me to retrieve accurate information about a dagrun for past runs, but it still doesn't solve the problem of having wrong information when a task is backfilled / cleared.
- wrapping every operator with a custom operator that exposes a custom `rendered_fields` that contains this information: this could be a nice hack, but it means I'll have to wrap every operator with this custom one.
- including the information as part of `executor_config`: this one could have been a nice solution since it is very similar to what I am looking for, the ability to pass down information to the task instance from the task, but it is a Python dict and cannot be JSON, which means when we get it from API other tools and technologies cannot parse it easily.

All in all, all of these approaches feel like hacks, therefore the ability to associate custom metadata with task instances, something that would be propagated from the task to the task instance, and from dag to dagrun, could solve this. 

I appreciate the brainstorming by the way, I am not saying this cannot be solved, but I am trying to outline what are the things that I thought of and didn't help so that it helps the conversation.

Have a great week!

**Reply by potiuk** [reply]
I think this is entirely unnecessary. https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_dag_run gives you "conf" - so you can already retrieve the configuraiton that you run your taks with, transitively -> task -> run_id -> dag_run. Is that not enough for you?

**Reply by potiuk** [reply]
BTW. DAG versioning is much more complex problem than just "configuration" - we have the Draft AIP https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-36+DAG+Versioning  prepered a long time ago - this is still a draft and requires a lot more thinking and handling many more cases and scenarios. Hopefully soon we will address it as some other pressing issues have been already handled, and I think we are slowly getting to the point that we can colllectively as community focus on AIP-36, but this will be a  HUGE effort - at least  few months of finalizing (or actually even restarting) discussion on that, and a LOT of changes touching nearly all aspects of airflow (database, scheduler, UI, API , DAG File Processor) and trying to handle just one small aspect of it separately without having at least a clear vision where we want to go with DAG versioning is kinda useless.

I think in your case basing on APIs and mapping Task-> Configuration (and potentially handling any queries like "which tasks are run with this version of configuration") outside of Airflow's core functionality. And discussing DAG any related changes as part of DAG versioning when we get to it, is likely going to be much more productive/constructive

**Reply by karakanb** [reply]
It is a very complex problem indeed, point taken. I will also go through the proposal.

About the `conf`, I agree that it would solve part of my problem, but I couldn't find a way of generating it at DAG generation time. Do you have an example where I can define the config as part of the DAG creation, and it would be passed onto the dagruns like you mentioned?  Would `params` do that?

**Reply by karakanb** [reply]
I tried passing it as part of `params`, and even though they are passed down to the task instance, the `conf` still returns empty, and the `/dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}` endpoint doesn't return the params given to the instance. Did I misunderstand what you suggested?

**Reply by potiuk** [reply]
No - you cant add it at the DAG generation time. THat makes completely no sense unless you had trouble with explaining what you want to do. 

I believe you wanted to see the conf that was used with a give dag_run not Dag. Why do you need configuration on DAG creation time?  

Don't you want to see conf that has been used during the DAG triggering ? Maybe you get it empty because you do NOT pass it when you trigger the dags (you should).

How do you want to distinguish different dag runs if you do not pass configuration when you trigger the dag run? 

Yes. You can trigger run with configuration via UI and check if you passed parameters configuration is there. And you - ideally need to define params so that dag triggering can validate them when you pass them.

If you do not see conf for the dag_run that was properly  triggered with parameters, maybe that is a bug in the version you are using - ideally upgrade to latest or try to find some information in the changelog about it.

**Reply by karakanb** [reply]
I guess I failed to explain what I wanted to do. I don't trigger dags manually, I let Airflow do all that, I just need to know what was the config that got used to generate that DAG structure, that's all.

---

Take a look at the following configuration:
```yaml
## let's call this "config V2"
id: my-dag
schedule: daily
tasks:
  - id: task1
    type: bash
    says: "task1 v1"
```

Assume I generate a DAG with a daily schedule from this configuration dynamically, which would include 1 BashOperator task called `task1` that prints out `task1 v1`. Now, when this DAG is scheduled automatically:
- since it has "daily" schedule, it will generate a dagrun, let's say with ID `scheduled__dagrun1`, 
- it will contain an instance of `task1`
- this instance of `task1` will print "task1 v1" via the BashOperator. 

All good so far, regular dynamic pipeline generation.

---

Now assume, the next day, the configuration is updated to be like this:
```yaml
## let's call this "config V2"
id: my-dag
schedule: daily
tasks:
  - id: task1
    type: bash
    says: "task1 v2"
```

With this configuration, the new DAG structure will be pretty much the same, except for what `task1` prints. Now, when this DAG is scheduled automatically:
- since it has "daily" schedule, it will generate a new dagrun, let's say with ID `scheduled__dagrun2`, 
- it will contain an instance of `task1`
- this time, the instance of `task1` will print "task1 v2" via the BashOperator. 

---

Now, what I want to do is to somehow get the association between `scheduled__dagrun2` and "config V2" contents, and do the same for `scheduled__dagrun1` and "config V1". 

For example, if I was able to add some metadata to the tasks/DAGs during the generation time that would get passed down to the dagruns, such as `default_args`, I could embed the YAML content as a dictionary there, and get them from the API, which means when I get the DAG details for `scheduled__dagrun1`, I would see that the config that was used to generate that DAG structure that got reflected in that run would be "config V1".

Does this make sense?

**Reply by potiuk** [reply]
Yes. And this is impossible now (literally), and this is what DAGVersioning AIP is supposed to target. There is no way of linking the "source" of dag generation with generated dag - simply because there is only "latest" DAG. We keep absolutely no information about when, how and by whom the DAG is generated. Currently scheduler only "overrides" the last version of DAG with one that is result of latest parsing - and does not store ANYTHING for ANY past version. The only things that are stored are actual DAG Runs that executed - but this is by-product of executing them. 

DAG versioning AIP is supposed to target exactly that - store historical versions of DAGs, match them with dag-runs that were used while given "structure" (which also matches the current "configuration") that was used then and be able to reflect it in the UI, database. 

So what you want to achieve is not posible now, simply because Airflow does not yet have this AIP implemented.

**Reply by karakanb** [reply]
Alright then, thank you very much for the discussion and input, I'll close this discussion for now.

**Comment by karakanb** [comment, marked as answer]
This doesn't seem to be possible as it stands today, please see here: https://github.com/apache/airflow/discussions/26462#discussioncomment-3703388
